session-manager/src/session_context.rs

```diff
@@ -130,13 +139,11 @@ impl SessionContext {
         server_config.ciphersuites = vec![supported_ciphersuite];
         server_config.versions = vec![rustls::ProtocolVersion::TLSv1_2];
 
-        Ok(Self {
-            server_config,
-            principals,
-            policy,
-            server_public_key,
-            server_private_key,
-        })
+        self.server_config = Some(server_config);
+        self.principals = Some(principals);
+        self.policy = Some(policy);
+
+        return Ok(());
     }
 
     pub fn set_cert_chain(&mut self, chain_data: &Vec<Vec<u8>>) -> Result<(), SessionManagerError> {
```

Comment:
Shouldn't this function take ownership of `chain_data` rather than borrowing it and cloning?

Reply:
I'm curious on this. 

I've read that we should be biased to receiving values by reference so we don't copy stuff around. And in general, taking ownership unnecessarily is bad, because it ties the hands of the caller (they will need to do their own copy on the call if they want to use it after the call). 

I don't really know what the right thing to do in these situations is.

Reply:
Passing by value means one less allocation if the caller has a copy that they are never going to use again.  Generally, I think functions that "stash" a value internally in some data structure take the thing they are going to stash by value, or at least that's the pattern followed in the standard library itself, e.g. in `std::vec::Vec` and `std::collections::HashMap`.

Though, I think this is a point on which reasonable minds can differ, so let's just leave it as-is...